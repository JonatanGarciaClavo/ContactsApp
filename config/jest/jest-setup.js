/*  eslint-disable */
// Make Enzyme functions available in all test files without importing
import { shallow, render, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
global.shallow = shallow;
global.render = render;
global.render = mount;
global.shallowToJson = shallowToJson;

// Skip createElement warnings but fail tests on any other warning
console.error = message => {
  if (!/(React.createElement: type should not be null)/.test(message)) {
      throw new Error(message);
  }
};