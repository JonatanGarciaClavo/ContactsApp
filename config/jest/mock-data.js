export const contact = {
  id: '-1',
  email: 'test@test.com',
  imgUrl: 'http://www.test.image.com',
  name: 'Tester',
  phoneNumber: '666666666',
};

export const contacts = [
  {
    id: '-1',
    email: 'test@test.com',
    imgUrl: 'http://www.test.image.com',
    name: 'Tester',
    phoneNumber: '666666666',
  },
  {
    id: '-2',
    email: 'test2@test.com',
    imgUrl: 'http://www.test2.image.com',
    name: 'Tester2',
    phoneNumber: '766666666',
  },
];

export const id = '-1';
