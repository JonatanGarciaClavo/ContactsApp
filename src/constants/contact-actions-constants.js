export const LOADING_CONTACT = 'LOADING_CONTACT';
export const INITIALIZE_CONTACT = 'INITIALIZE_CONTACT';
export const UPDATE_CONTACT_ATTRIBUTE = 'UPDATE_CONTACT_ATTRIBUTE';
export const VALIDATE_CONTACT_ATTRIBUTE = 'VALIDATE_CONTACT_ATTRIBUTE';
export const VALIDATE_CONTACT = 'VALIDATE_CONTACT';
export const REQUEST_CONTACT = 'REQUEST_CONTACT';
export const REQUEST_SAVE_CONTACT = 'REQUEST_SAVE_CONTACT';
