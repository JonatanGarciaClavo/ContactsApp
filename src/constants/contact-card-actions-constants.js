export const LOADING_CONTACT_CARD = 'LOADING_CONTACT_CARD';
export const INITIALIZE_CONTACT_CARD = 'INITIALIZE_CONTACT_CARD';
export const REQUEST_CONTACT_CARD = 'REQUEST_CONTACT_CARD';
export const REQUEST_DELETE_CONTACT_CARD = 'REQUEST_DELETE_CONTACT_CARD';
export const INITILIZE_CONTACT_CARD_FROM_OTHER_VIEW = 'INITILIZE_CONTACT_CARD_FROM_OTHER_VIEW';
