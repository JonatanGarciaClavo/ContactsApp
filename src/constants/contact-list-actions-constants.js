export const LOADING_CONTACT_LIST = 'LOADING_CONTACT_LIST';
export const INITIALIZE_CONTACT_LIST = 'INITIALIZE_CONTACT_LIST';
export const CHANGE_LIST_MODE = 'CHANGE_LIST_MODE';
