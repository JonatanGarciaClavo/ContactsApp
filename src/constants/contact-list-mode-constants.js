export const LIST_MODE = 'list';
export const CARD_MODE = 'card'
