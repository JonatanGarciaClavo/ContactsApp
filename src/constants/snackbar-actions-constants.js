export const CLOSE_ERROR_MESSAGE = 'CLOSE_ERROR_MESSAGE';
export const SET_ERROR_MESSAGE = 'SET_ERROR_MESSAGE';
