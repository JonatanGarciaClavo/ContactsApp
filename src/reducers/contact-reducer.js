import _ from 'lodash';
import { Record, Map, fromJS } from 'immutable';
import { REQUEST_CONTACT, REQUEST_SAVE_CONTACT, REQUEST_CONTACT_SUCCESS, UPDATE_CONTACT_ATTRIBUTE,
  VALIDATE_CONTACT, VALIDATE_CONTACT_ATTRIBUTE, RESET_CONTACT,
  } from '../constants/contact-actions-constants';
import { SET_ERROR_MESSAGE } from '../constants/snackbar-actions-constants';
import ContactModel from '../models/ContactModel';

const requiredValidation = value =>
  value !== undefined && value !== null && (!_.isEmpty(value) || _.isNumber(value));

const isEmail = (value) => {
  if (value) {
    const regExp = /^((([a-z]|\d|[!#\$%&'\*\+\-\/=\?\^_`{\|}~]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+(\.([a-z]|\d|[!#\$%&'\*\+\-\/=\?\^_`{\|}~]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+)*)|((\x22)((((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(([\x01-\x08\x0b\x0c\x0e-\x1f\x7f]|\x21|[\x23-\x5b]|[\x5d-\x7e]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(\\([\x01-\x09\x0b\x0c\x0d-\x7f]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]))))*(((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(\x22)))@((([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))\.)+(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))$/i;// eslint-disable-line
    return regExp.test(value);
  }
  return true;
};

const isUrl = (value) => {
  if (value) {
    const regExp = new RegExp("^(?:(?:https?|ftp):\/\/)(?:\S+(?::\S*)?@)?(?:(?!10(?:\.\d{1,3}){3})(?!127(?:\.\d{1,3}){3})(?!169\.254(?:\.\d{1,3}){2})(?!192\.168(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\u00a1-\uffff0-9]+-?)*[a-z\u00a1-\uffff0-9]+)(?:\.(?:[a-z\u00a1-\uffff0-9]+-?)*[a-z\u00a1-\uffff0-9]+)*(?:\.(?:[a-z\u00a1-\uffff]{2,})))(?::\d{2,5})?(?:\/[^\s]*)?");// eslint-disable-line
    return regExp.test(value);
  }
  return true;
};

const applyValidator = (ruleObject, value) => {
  let error;
  let functionValidator;
  if (_.isFunction(ruleObject.rule)) {
    functionValidator = ruleObject.rule;
  }
  if (functionValidator && !functionValidator(value)) {
    error = ruleObject.error || `${ruleObject.rule} not followed`;
  } else if (!functionValidator) {
    console.warn('applyValidator: %s does not exists', ruleObject.rule);// eslint-disable-line
  }
  return error;
};

const validateAll = (rulesObject, contact) => {
  const errors = {};

  _.each(rulesObject, (rules, name) => {
    if (rules && _.isArray(rules)) {
      const value = contact[name];
      _.each(rules, (ruleObject) => {
        const error = applyValidator(ruleObject, value);
        if (error) {
          errors[name] = error;
        }
      });
    } else {
      console.warn('validateAll: missing parameters rules ');// eslint-disable-line
    }
  });

  return _.isEmpty(errors) ? false : errors;
};

const rulesObject = {
  name: [{ rule: requiredValidation, error: 'Name is required' }],
  email: [
    { rule: requiredValidation, error: 'Email is required' },
    { rule: isEmail, error: 'Email format is incorrect' },
  ],
  imgUrl: [{ rule: isUrl, error: 'Invalid url' }],
};

export const Contact = new Record({
  loading: false,
  isModified: false,
  contact: new ContactModel(),
  errors: new Map(),
});


const contact = (state = new Contact(), action) => {
  const { name, value } = action;
  let error;
  let errors;
  switch (action.type) {
    case (REQUEST_SAVE_CONTACT):
    case (REQUEST_CONTACT):
      return state.set('loading', true);
    case (SET_ERROR_MESSAGE):
      return state.set('loading', false);
    case (REQUEST_CONTACT_SUCCESS):
      return state
        .set('contact', action.contact || new ContactModel())
        .set('loading', false)
        .set('errors', new Map());
    case (UPDATE_CONTACT_ATTRIBUTE):
      return state
        .set('isModified', true)
        .setIn(['contact', name], value);
    case VALIDATE_CONTACT_ATTRIBUTE:
      error = validateAll(_.pick(rulesObject, name), { [name]: value });
      if (error && error[name]) {
        return state.setIn(['errors', name], error[name]);
      }
      return state.deleteIn(['errors', name]);
    case VALIDATE_CONTACT:
      errors = validateAll(rulesObject, state.get('contact').toJS());
      return state.set('errors', new Map(fromJS(errors || {})));
    case RESET_CONTACT:
      return new Contact();
    default:
      return state;
  }
};

export default contact;
