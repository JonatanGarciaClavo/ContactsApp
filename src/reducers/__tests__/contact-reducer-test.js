import Immutable from 'immutable';
import { initialState, default as contact } from '../contact-reducer';
import { REQUEST_CONTACT, REQUEST_SAVE_CONTACT, REQUEST_CONTACT_SUCCESS, UPDATE_CONTACT_ATTRIBUTE,
  VALIDATE_CONTACT, VALIDATE_CONTACT_ATTRIBUTE, RESET_CONTACT,
  } from '../../constants/contact-actions-constants';
import { SET_ERROR_MESSAGE } from '../../constants/snackbar-actions-constants';
import { contact as mockContact } from '../../../config/jest/mock-data';

const testContact = new Immutable.Map(Immutable.fromJS(mockContact));

describe('Test contact reducer', () => {
  it('returns the initialState state on an undefined state', () => {
    expect(contact(undefined, { type: '_NULL' })).toMatchSnapshot();
  });
  it('returns the same state on an unhandled action', () => {
    expect(contact(initialState, { type: '_NULL' })).toMatchSnapshot();
  });

  it('handles REQUEST_CONTACT action', () => {
    expect(contact(initialState, { type: REQUEST_CONTACT })).toMatchSnapshot();
  });
  it('handles REQUEST_SAVE_CONTACT action', () => {
    expect(contact(initialState, { type: REQUEST_SAVE_CONTACT })).toMatchSnapshot();
  });
  it('handles SET_ERROR_MESSAGE action', () => {
    expect(contact(initialState, { type: SET_ERROR_MESSAGE })).toMatchSnapshot();
  });

  it('handles REQUEST_CONTACT_SUCCESS action', () => {
    expect(contact(initialState, { type: REQUEST_CONTACT_SUCCESS, contact: testContact }))
      .toMatchSnapshot();
  });

  it('handles UPDATE_CONTACT_ATTRIBUTE action', () => {
    const newState = contact(initialState, { type: REQUEST_CONTACT_SUCCESS, contact: testContact });
    expect(contact(newState, {
      type: UPDATE_CONTACT_ATTRIBUTE,
      name: 'email',
      value: 'test2@test.com',
    })).toMatchSnapshot();
    expect(contact(newState, {
      type: UPDATE_CONTACT_ATTRIBUTE,
      name: 'imgUrl',
      value: 'http://www.test.image.com/2.png',
    })).toMatchSnapshot();
    expect(contact(newState, {
      type: UPDATE_CONTACT_ATTRIBUTE,
      name: 'name',
      value: 'Tester pro',
    })).toMatchSnapshot();
    expect(contact(newState, {
      type: UPDATE_CONTACT_ATTRIBUTE,
      name: 'phoneNumber',
      value: '777777777',
    })).toMatchSnapshot();
  });

  it('handles VALIDATE_CONTACT_ATTRIBUTE email action', () => {
    expect(contact(initialState, {
      type: VALIDATE_CONTACT_ATTRIBUTE,
      name: 'email',
      value: 'test@test.com',
    })).toMatchSnapshot();
    expect(contact(initialState, {
      type: VALIDATE_CONTACT_ATTRIBUTE,
      name: 'email',
      value: '',
    })).toMatchSnapshot();
    expect(contact(initialState, {
      type: VALIDATE_CONTACT_ATTRIBUTE,
      name: 'email',
      value: 'test@test',
    })).toMatchSnapshot();
  });

  it('handles VALIDATE_CONTACT_ATTRIBUTE imgUrl action', () => {
    expect(contact(initialState, {
      type: VALIDATE_CONTACT_ATTRIBUTE,
      name: 'imgUrl',
      value: 'http://www.test.image.com',
    })).toMatchSnapshot();
    expect(contact(initialState, {
      type: VALIDATE_CONTACT_ATTRIBUTE,
      name: 'imgUrl',
      value: 'invalidURL',
    })).toMatchSnapshot();
  });

  it('handles VALIDATE_CONTACT_ATTRIBUTE name action', () => {
    expect(contact(initialState, {
      type: VALIDATE_CONTACT_ATTRIBUTE,
      name: 'name',
      value: 'Tester',
    })).toMatchSnapshot();
    expect(contact(initialState, {
      type: VALIDATE_CONTACT_ATTRIBUTE,
      name: 'name',
      value: '',
    })).toMatchSnapshot();
  });

  it('handles VALIDATE_CONTACT success action', () => {
    const newState = contact(initialState, { type: REQUEST_CONTACT_SUCCESS, contact: testContact });
    expect(contact(newState, { type: VALIDATE_CONTACT })).toMatchSnapshot();
  });

  it('handles RESET_CONTACT action', () => {
    const newState = contact(initialState, { type: REQUEST_CONTACT_SUCCESS, contact: testContact });
    expect(contact(newState, { type: RESET_CONTACT }))
      .toMatchSnapshot();
  });
});
