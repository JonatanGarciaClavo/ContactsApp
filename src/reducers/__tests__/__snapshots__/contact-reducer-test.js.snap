exports[`Test contact reducer handles INITIALIZE_CONTACT action 1`] = `
Object {
  "contact": Object {
    "email": "test@test.com",
    "imgUrl": "http://www.test.image.com",
    "name": "Tester",
    "phoneNumber": "666666666",
  },
  "errors": Object {},
  "isModified": false,
  "loading": false,
}
`;

exports[`Test contact reducer handles LOADING_CONTACT action 1`] = `
Object {
  "contact": Object {
    "email": "",
    "imgUrl": "",
    "name": "",
    "phoneNumber": "",
  },
  "errors": Object {},
  "isModified": false,
  "loading": true,
}
`;

exports[`Test contact reducer handles UPDATE_CONTACT_ATTRIBUTE action 1`] = `
Object {
  "contact": Object {
    "email": "test2@test.com",
    "imgUrl": "http://www.test.image.com",
    "name": "Tester",
    "phoneNumber": "666666666",
  },
  "errors": Object {},
  "isModified": true,
  "loading": false,
}
`;

exports[`Test contact reducer handles UPDATE_CONTACT_ATTRIBUTE action 2`] = `
Object {
  "contact": Object {
    "email": "test@test.com",
    "imgUrl": "http://www.test.image.com/2.png",
    "name": "Tester",
    "phoneNumber": "666666666",
  },
  "errors": Object {},
  "isModified": true,
  "loading": false,
}
`;

exports[`Test contact reducer handles UPDATE_CONTACT_ATTRIBUTE action 3`] = `
Object {
  "contact": Object {
    "email": "test@test.com",
    "imgUrl": "http://www.test.image.com",
    "name": "Tester pro",
    "phoneNumber": "666666666",
  },
  "errors": Object {},
  "isModified": true,
  "loading": false,
}
`;

exports[`Test contact reducer handles UPDATE_CONTACT_ATTRIBUTE action 4`] = `
Object {
  "contact": Object {
    "email": "test@test.com",
    "imgUrl": "http://www.test.image.com",
    "name": "Tester",
    "phoneNumber": "777777777",
  },
  "errors": Object {},
  "isModified": true,
  "loading": false,
}
`;

exports[`Test contact reducer handles VALIDATE_CONTACT success action 1`] = `
Object {
  "contact": Object {
    "email": "",
    "imgUrl": "",
    "name": "",
    "phoneNumber": "",
  },
  "errors": Object {
    "email": "Email is required",
    "name": "Name is required",
  },
  "isModified": false,
  "loading": false,
}
`;

exports[`Test contact reducer handles VALIDATE_CONTACT_ATTRIBUTE email action 1`] = `
Object {
  "contact": Object {
    "email": "",
    "imgUrl": "",
    "name": "",
    "phoneNumber": "",
  },
  "errors": Object {},
  "isModified": false,
  "loading": false,
}
`;

exports[`Test contact reducer handles VALIDATE_CONTACT_ATTRIBUTE email action 2`] = `
Object {
  "contact": Object {
    "email": "",
    "imgUrl": "",
    "name": "",
    "phoneNumber": "",
  },
  "errors": Object {
    "email": "Email is required",
  },
  "isModified": false,
  "loading": false,
}
`;

exports[`Test contact reducer handles VALIDATE_CONTACT_ATTRIBUTE email action 3`] = `
Object {
  "contact": Object {
    "email": "",
    "imgUrl": "",
    "name": "",
    "phoneNumber": "",
  },
  "errors": Object {
    "email": "Email format is incorrect",
  },
  "isModified": false,
  "loading": false,
}
`;

exports[`Test contact reducer handles VALIDATE_CONTACT_ATTRIBUTE imgUrl action 1`] = `
Object {
  "contact": Object {
    "email": "",
    "imgUrl": "",
    "name": "",
    "phoneNumber": "",
  },
  "errors": Object {},
  "isModified": false,
  "loading": false,
}
`;

exports[`Test contact reducer handles VALIDATE_CONTACT_ATTRIBUTE imgUrl action 2`] = `
Object {
  "contact": Object {
    "email": "",
    "imgUrl": "",
    "name": "",
    "phoneNumber": "",
  },
  "errors": Object {
    "imgUrl": "Invalid url",
  },
  "isModified": false,
  "loading": false,
}
`;

exports[`Test contact reducer handles VALIDATE_CONTACT_ATTRIBUTE name action 1`] = `
Object {
  "contact": Object {
    "email": "",
    "imgUrl": "",
    "name": "",
    "phoneNumber": "",
  },
  "errors": Object {},
  "isModified": false,
  "loading": false,
}
`;

exports[`Test contact reducer handles VALIDATE_CONTACT_ATTRIBUTE name action 2`] = `
Object {
  "contact": Object {
    "email": "",
    "imgUrl": "",
    "name": "",
    "phoneNumber": "",
  },
  "errors": Object {
    "name": "Name is required",
  },
  "isModified": false,
  "loading": false,
}
`;

exports[`Test contact reducer returns the initialState state on an undefined state 1`] = `
Object {
  "contact": Object {
    "email": "",
    "imgUrl": "",
    "name": "",
    "phoneNumber": "",
  },
  "errors": Object {},
  "isModified": false,
  "loading": false,
}
`;

exports[`Test contact reducer returns the same state on an unhandled action 1`] = `
Object {
  "contact": Object {
    "email": "",
    "imgUrl": "",
    "name": "",
    "phoneNumber": "",
  },
  "errors": Object {},
  "isModified": false,
  "loading": false,
}
`;
